<!DOCTYPE html>

<html class="writer-html5" lang="en">
<head>
<meta charset="utf-8"/>
<meta content="IE=edge" http-equiv="X-UA-Compatible"/>
<meta content="width=device-width, initial-scale=1.0" name="viewport"/><link href="https://molgenis.github.io/vip/advanced/annotations/" rel="canonical"/>
<link href="../../img/favicon.ico" rel="shortcut icon"/>
<title>Annotations - Variant Interpretation Pipeline</title>
<link href="../../css/theme.css" rel="stylesheet"/>
<link href="../../css/theme_extra.css" rel="stylesheet"/>
<link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" rel="stylesheet"/>
<link href="../../custom.css" rel="stylesheet"/>
<link href="../../assets/stylesheets/panzoom.css" rel="stylesheet"/>
<script>
        // Current page data
        var mkdocs_page_name = "Annotations";
        var mkdocs_page_input_path = "advanced/annotations.md";
        var mkdocs_page_url = "/vip/advanced/annotations/";
      </script>
<!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]-->
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
<script>hljs.highlightAll();</script>
<meta name="panzoom-theme" content="readthedocs">
<meta name="panzoom-data" content='{"selectors": [".d2", ".mermaid"], "initial_zoom_level": 1.0}'> </head>
<body class="wy-body-for-nav" role="document">
<div class="wy-grid-for-nav">
<nav class="wy-nav-side stickynav" data-toggle="wy-nav-shift">
<div class="wy-side-scroll">
<div class="wy-side-nav-search">
<a class="icon icon-home" href="../.."> Variant Interpretation Pipeline
        </a><div role="search">
<form action="../../search.html" class="wy-form" id="rtd-search-form" method="get">
<input aria-label="Search docs" name="q" placeholder="Search docs" title="Type search term here" type="text"/>
</form>
</div>
</div>
<div aria-label="Navigation menu" class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation">
<p class="caption"><span class="caption-text">Home</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../..">Introduction</a>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../home/key_features/">Key features</a>
</li>
</ul>
<p class="caption"><span class="caption-text">Get Started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../get_started/requirements/">Requirements</a>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../get_started/installation/">Installation</a>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../get_started/start_running/">Start running</a>
</li>
</ul>
<p class="caption"><span class="caption-text">Usage</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../usage/command-line-options/">Command-line options</a>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../usage/workflow/">Workflow</a>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../usage/input/">Input</a>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../usage/output/">Output</a>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../usage/config/">Config</a>
</li>
</ul>
<p class="caption"><span class="caption-text">Examples</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../examples/nanopore/">Nanopore</a>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../examples/multi-project/">Multi-project</a>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../examples/reanalysis/">Reanalysis</a>
</li>
</ul>
<p class="caption"><span class="caption-text">Advanced</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal current" href="#">Annotations</a>
<ul class="current">
<li class="toctree-l2"><a class="reference internal" href="#overview">Overview</a>
<ul>
<li class="toctree-l3"><a class="reference internal" href="#filter">Filter</a>
</li>
<li class="toctree-l3"><a class="reference internal" href="#format">Format</a>
</li>
<li class="toctree-l3"><a class="reference internal" href="#info">INFO</a>
<ul>
<li class="toctree-l4"><a class="reference internal" href="#infocsq">INFO/CSQ</a>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#details">Details</a>
<ul>
<li class="toctree-l3"><a class="reference internal" href="#plugins">Plugins</a>
<ul>
<li class="toctree-l4"><a class="reference internal" href="#capice">CAPICE</a>
</li>
<li class="toctree-l4"><a class="reference internal" href="#vkgl">VKGL</a>
</li>
<li class="toctree-l4"><a class="reference internal" href="#spliceai">SpliceAI</a>
</li>
<li class="toctree-l4"><a class="reference internal" href="#annotsv">AnnotSV</a>
</li>
<li class="toctree-l4"><a class="reference internal" href="#hpo">HPO</a>
</li>
<li class="toctree-l4"><a class="reference internal" href="#inheritance">Inheritance</a>
</li>
<li class="toctree-l4"><a class="reference internal" href="#grantham">Grantham</a>
</li>
<li class="toctree-l4"><a class="reference internal" href="#gado">GADO</a>
</li>
<li class="toctree-l4"><a class="reference internal" href="#alphscore">AlphScore</a>
</li>
<li class="toctree-l4"><a class="reference internal" href="#ncer">ncER</a>
</li>
<li class="toctree-l4"><a class="reference internal" href="#remm">ReMM</a>
</li>
<li class="toctree-l4"><a class="reference internal" href="#fathmm-mkl">FATHMM-MKL</a>
</li>
<li class="toctree-l4"><a class="reference internal" href="#green-db-constraint-scores">GREEN-DB constraint scores</a>
</li>
<li class="toctree-l4"><a class="reference internal" href="#apogee-2">APOGEE 2</a>
</li>
<li class="toctree-l4"><a class="reference internal" href="#mitotip">MitoTIP</a>
</li>
<li class="toctree-l4"><a class="reference internal" href="#hmtvar">HmtVar</a>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../classification_trees/">Classification trees</a>
</li>
<li class="toctree-l1"><a class="reference internal" href="../report_templates/">Report templates</a>
</li>
</ul>
<p class="caption"><span class="caption-text">Help</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../help/issues/">Issues</a>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../help/frequently_asked_questions/">Frequently asked questions</a>
</li>
</ul>
<p class="caption"><span class="caption-text">About</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../about/license/">License</a>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../about/acknowledgements/">Acknowledgements</a>
</li>
</ul>
</div>
</div>
</nav>
<section class="wy-nav-content-wrap" data-toggle="wy-nav-shift">
<nav aria-label="Mobile navigation menu" class="wy-nav-top" role="navigation">
<i class="fa fa-bars" data-toggle="wy-nav-top"></i>
<a href="../..">Variant Interpretation Pipeline</a>
</nav>
<div class="wy-nav-content">
<div class="rst-content"><div aria-label="breadcrumbs navigation" role="navigation">
<ul class="wy-breadcrumbs">
<li><a aria-label="Docs" class="icon icon-home" href="../.."></a></li>
<li class="breadcrumb-item">Advanced</li>
<li class="breadcrumb-item active">Annotations</li>
<li class="wy-breadcrumbs-aside">
<a class="icon icon-github" href="https://github.com/molgenis/vip/edit/main/docs/advanced/annotations.md"> Edit on GitHub</a>
</li>
</ul>
<hr/>
</div>
<div class="document" itemscope="itemscope" itemtype="http://schema.org/Article" role="main">
<div class="section" itemprop="articleBody">
<h1 id="annotations">Annotations<a class="headerlink" href="#annotations" title="Permanent link">¶</a></h1>
<p>VIP annotates variant effects and genotype data for samples using a rich set of tools. Annotions can be used
to <a href="../classification_trees/">classify variants using classification trees</a> and displayed
in <a href="../report_templates/">reports</a>.</p>
<h2 id="overview">Overview<a class="headerlink" href="#overview" title="Permanent link">¶</a></h2>
<p>The table contains annotations available in most output files. Depending on the workflow and the configuration used
additional annotations might be available,
check the output file headers for the complete overview. Similarly, some annotations listed below might be missing from
your output file depending on the sample sheet content and configuration.</p>
<h3 id="filter">Filter<a class="headerlink" href="#filter" title="Permanent link">¶</a></h3>
<table>
<thead>
<tr>
<th>annotation</th>
<th>value</th>
<th>source</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td>FILTER</td>
<td>MONOALLELIC</td>
<td><a href="https://github.com/dnanexus-rnd/GLnexus">GLnexus</a></td>
<td>Site represents one ALT allele in a region with multiple variants that could not be unified into non-overlapping multi-allelic sites</td>
</tr>
<tr>
<td>FILTER</td>
<td>PASS</td>
<td><a href="https://github.com/molgenis/straglr-tsv2vcf">Straglr</a></td>
<td>All filters passed</td>
</tr>
<tr>
<td>FILTER</td>
<td>q5</td>
<td><a href="https://github.com/tjiangHIT/cuteSV">cuteSV</a></td>
<td>Quality below 5</td>
</tr>
<tr>
<td>FILTER</td>
<td>UNRESOLVED</td>
<td><a href="https://github.com/fritzsedlazeck/Spectre">Spectre</a></td>
<td>An insertion that is longer than the read and thus we cannot predict the full size</td>
</tr>
</tbody>
</table>
<h3 id="format">Format<a class="headerlink" href="#format" title="Permanent link">¶</a></h3>
<table>
<thead>
<tr>
<th>annotation</th>
<th>type</th>
<th>source</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td>FORMAT/CD</td>
<td>float list</td>
<td><a href="https://github.com/fritzsedlazeck/Spectre">Spectre</a></td>
<td>Coverage read depth</td>
</tr>
<tr>
<td>FORMAT/DP</td>
<td>integer</td>
<td></td>
<td>Approximate read depth (reads with MQ=255 or with bad mates are filtered)</td>
</tr>
<tr>
<td>FORMAT/DR</td>
<td>integer</td>
<td><a href="https://github.com/tjiangHIT/cuteSV">cuteSV</a></td>
<td># High-quality reference reads</td>
</tr>
<tr>
<td>FORMAT/DV</td>
<td>integer</td>
<td><a href="https://github.com/tjiangHIT/cuteSV">cuteSV</a></td>
<td># High-quality variant reads</td>
</tr>
<tr>
<td>FORMAT/GQ</td>
<td>integer</td>
<td><a href="https://github.com/tjiangHIT/cuteSV">cuteSV</a>, <a href="https://github.com/fritzsedlazeck/Spectre">Spectre</a></td>
<td>Genotype Quality</td>
</tr>
<tr>
<td>FORMAT/GT</td>
<td>string</td>
<td><a href="https://github.com/tjiangHIT/cuteSV">cuteSV</a>, <a href="https://github.com/fritzsedlazeck/Spectre">Spectre</a>, <a href="https://github.com/molgenis/straglr-tsv2vcf">straglr-tsv2vcf</a></td>
<td>Genotype</td>
</tr>
<tr>
<td>FORMAT/HO</td>
<td>float list</td>
<td><a href="https://github.com/fritzsedlazeck/Spectre">Spectre</a></td>
<td>Homozygosity proportion</td>
</tr>
<tr>
<td>FORMAT/ID</td>
<td>string</td>
<td><a href="https://github.com/fritzsedlazeck/Spectre">Spectre</a></td>
<td>Population ID of supporting CNV calls</td>
</tr>
<tr>
<td>FORMAT/LC</td>
<td>integer</td>
<td><a href="https://github.com/molgenis/straglr-tsv2vcf">straglr-tsv2vcf</a></td>
<td>Locus coverage</td>
</tr>
<tr>
<td>FORMAT/PL</td>
<td>integer list</td>
<td><a href="https://github.com/tjiangHIT/cuteSV">cuteSV</a></td>
<td># Phred-scaled genotype likelihoods rounded to the closest integer</td>
</tr>
<tr>
<td>FORMAT/PS</td>
<td>integer</td>
<td>Whatshap</td>
<td>Phase set identifier</td>
</tr>
<tr>
<td>FORMAT/RNC</td>
<td>character list</td>
<td></td>
<td>Reason for No Call in GT: . = n/a, M = Missing data, P = Partial data, I = gVCF input site is non-called, D = insufficient Depth of coverage, - = unrepresentable overlapping deletion, L = Lost/unrepresentable allele (other than deletion), U = multiple Unphased variants present, O = multiple Overlapping variants present, 1 = site is Monoallelic, no assertion about presence of REF or ALT allele</td>
</tr>
<tr>
<td>FORMAT/RU_CALL</td>
<td>string</td>
<td><a href="https://github.com/molgenis/straglr-tsv2vcf">straglr-tsv2vcf</a></td>
<td>Most frequent actual repeat motif</td>
</tr>
<tr>
<td>FORMAT/RU_CI</td>
<td>string</td>
<td><a href="https://github.com/molgenis/straglr-tsv2vcf">straglr-tsv2vcf</a></td>
<td>95% confidence interval per allele. 'NA' if less than 2 reads were present.</td>
</tr>
<tr>
<td>FORMAT/RU_MATCH</td>
<td>integer</td>
<td><a href="https://github.com/molgenis/straglr-tsv2vcf">straglr-tsv2vcf</a></td>
<td>RU call matches catalog (allowing shift/IUPAC, 1= match, 0 = mismatch</td>
</tr>
<tr>
<td>FORMAT/RU_NR</td>
<td>string</td>
<td><a href="https://github.com/molgenis/straglr-tsv2vcf">straglr-tsv2vcf</a></td>
<td>Number of repeat units per allele.</td>
</tr>
<tr>
<td>FORMAT/RU_SEEN</td>
<td>string</td>
<td><a href="https://github.com/molgenis/straglr-tsv2vcf">straglr-tsv2vcf</a></td>
<td>All RUs encountered with read counts</td>
</tr>
<tr>
<td>FORMAT/RU_SPAN</td>
<td>integer</td>
<td><a href="https://github.com/molgenis/straglr-tsv2vcf">straglr-tsv2vcf</a></td>
<td>Number of spanning reads per allele.</td>
</tr>
<tr>
<td>FORMAT/VI</td>
<td>string list</td>
<td><a href="https://github.com/molgenis/vip-inheritance-matcher/">vip-inheritance-matcher</a></td>
<td>An enumeration of possible inheritance modes (Possible values: AR, AR_C, AD, AD_IP, XLR, XLD)</td>
</tr>
<tr>
<td>FORMAT/VIAB</td>
<td>float</td>
<td>annotate process</td>
<td>VIP calculated allele balance</td>
</tr>
<tr>
<td>FORMAT/VIC</td>
<td>string</td>
<td><a href="https://github.com/molgenis/vip-inheritance-matcher/">vip-inheritance-matcher</a></td>
<td>Possible Compound hetrozygote variants</td>
</tr>
<tr>
<td>FORMAT/VID</td>
<td>integer</td>
<td><a href="https://github.com/molgenis/vip-inheritance-matcher/">vip-inheritance-matcher</a></td>
<td>De novo variant</td>
</tr>
<tr>
<td>FORMAT/VIG</td>
<td>string list</td>
<td><a href="https://github.com/molgenis/vip-inheritance-matcher/">vip-inheritance-matcher</a></td>
<td>Genes with an inheritance match</td>
</tr>
<tr>
<td>FORMAT/VIM</td>
<td>integer</td>
<td><a href="https://github.com/molgenis/vip-inheritance-matcher/">vip-inheritance-matcher</a></td>
<td>Inheritance Match: Genotypes, affected statuses and known gene inheritance patterns match</td>
</tr>
<tr>
<td>FORMAT/VIPC_S</td>
<td>string list</td>
<td><a href="https://github.com/molgenis/vip-decision-tree/">vip-decision-tree</a></td>
<td>VIP decision tree classification for sample</td>
</tr>
<tr>
<td>FORMAT/VIPP_S</td>
<td>string list</td>
<td><a href="https://github.com/molgenis/vip-decision-tree/">vip-decision-tree</a></td>
<td>VIP decision tree path for sample</td>
</tr>
</tbody>
</table>
<h3 id="info">INFO<a class="headerlink" href="#info" title="Permanent link">¶</a></h3>
<table>
<thead>
<tr>
<th>annotation</th>
<th>type</th>
<th>source</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td>INFO/AC</td>
<td>integer list</td>
<td></td>
<td>Allele count in genotypes</td>
</tr>
<tr>
<td>INFO/AF</td>
<td>float list</td>
<td><a href="https://github.com/tjiangHIT/cuteSV">cuteSV</a></td>
<td>Allele Frequency</td>
</tr>
<tr>
<td>INFO/AN</td>
<td>integer</td>
<td></td>
<td>Total number of alleles in called genotypes</td>
</tr>
<tr>
<td>INFO/AQ</td>
<td>integer list</td>
<td></td>
<td>Allele Quality score reflecting evidence for each alternate allele (Phred scale)</td>
</tr>
<tr>
<td>INFO/CHR2</td>
<td>string</td>
<td><a href="https://github.com/tjiangHIT/cuteSV">cuteSV</a></td>
<td>Chromosome for END coordinate in case of a translocation</td>
</tr>
<tr>
<td>INFO/CILEN</td>
<td>integer list</td>
<td><a href="https://github.com/tjiangHIT/cuteSV">cuteSV</a></td>
<td>Confidence interval around inserted/deleted material between breakends</td>
</tr>
<tr>
<td>INFO/CIPOS</td>
<td>integer list</td>
<td><a href="https://github.com/tjiangHIT/cuteSV">cuteSV</a></td>
<td>Confidence interval around POS for imprecise variants</td>
</tr>
<tr>
<td>INFO/CN</td>
<td>integer</td>
<td><a href="https://github.com/fritzsedlazeck/Spectre">Spectre</a></td>
<td>Estimated copy number status</td>
</tr>
<tr>
<td>INFO/Disease</td>
<td>string</td>
<td><a href="https://github.com/Clinical-Genomics/stranger/">Stranger</a></td>
<td>Associated disorder</td>
</tr>
<tr>
<td>INFO/DisplayRU</td>
<td>string</td>
<td><a href="https://github.com/Clinical-Genomics/stranger/">Stranger</a></td>
<td>Display repeat unit familiar to clinician</td>
</tr>
<tr>
<td>INFO/END</td>
<td>integer</td>
<td><a href="https://github.com/tjiangHIT/cuteSV">cuteSV</a>, <a href="https://github.com/fritzsedlazeck/Spectre">Spectre</a>, <a href="https://github.com/molgenis/straglr-tsv2vcf">straglr-tsv2vcf</a></td>
<td>End position of the structural variant</td>
</tr>
<tr>
<td>INFO/HGNCId</td>
<td>integer</td>
<td><a href="https://github.com/Clinical-Genomics/stranger/">Stranger</a></td>
<td>HGNC gene id for associated disease gene</td>
</tr>
<tr>
<td>INFO/IMPRECISE</td>
<td>boolean</td>
<td><a href="https://github.com/tjiangHIT/cuteSV">cuteSV</a></td>
<td>Imprecise structural variant</td>
</tr>
<tr>
<td>INFO/InheritanceMode</td>
<td>string</td>
<td><a href="https://github.com/Clinical-Genomics/stranger/">Stranger</a></td>
<td>Main mode of inheritance for disorder</td>
</tr>
<tr>
<td>INFO/OLD_REC</td>
<td>string</td>
<td></td>
<td>Original variant. Format: CHR|POS|REF|ALT|USED_ALT_IDX</td>
</tr>
<tr>
<td>INFO/PRECISE</td>
<td>boolean</td>
<td><a href="https://github.com/tjiangHIT/cuteSV">cuteSV</a></td>
<td>Precise structural variant</td>
</tr>
<tr>
<td>INFO/RankScore</td>
<td>string</td>
<td><a href="https://github.com/Clinical-Genomics/stranger/">Stranger</a></td>
<td>RankScore for variant in this family as family(str):score(int)</td>
</tr>
<tr>
<td>INFO/RE</td>
<td>integer</td>
<td><a href="https://github.com/tjiangHIT/cuteSV">cuteSV</a></td>
<td>Number of read support this record</td>
</tr>
<tr>
<td>INFO/REPID</td>
<td>string</td>
<td><a href="https://github.com/molgenis/straglr-tsv2vcf">straglr-tsv2vcf</a></td>
<td>Locus identifier.</td>
</tr>
<tr>
<td>INFO/RU_CAT</td>
<td>string</td>
<td><a href="https://github.com/molgenis/straglr-tsv2vcf">straglr-tsv2vcf</a></td>
<td>Catalog repeat motif</td>
</tr>
<tr>
<td>INFO/Source</td>
<td>string</td>
<td><a href="https://github.com/Clinical-Genomics/stranger/">Stranger</a></td>
<td>Source collection for variant definition</td>
</tr>
<tr>
<td>INFO/SourceDisplay</td>
<td>string</td>
<td><a href="https://github.com/Clinical-Genomics/stranger/">Stranger</a></td>
<td>Source for variant definition, display</td>
</tr>
<tr>
<td>INFO/SourceId</td>
<td>string</td>
<td><a href="https://github.com/Clinical-Genomics/stranger/">Stranger</a></td>
<td>Source id for variant definition</td>
</tr>
<tr>
<td>INFO/STR_NORMAL_MAX</td>
<td>integer</td>
<td><a href="https://github.com/Clinical-Genomics/stranger/">Stranger</a></td>
<td>Max number of repeats allowed to call as normal</td>
</tr>
<tr>
<td>INFO/STR_PATHOLOGIC_MIN</td>
<td>integer</td>
<td><a href="https://github.com/Clinical-Genomics/stranger/">Stranger</a></td>
<td>Min number of repeats required to call as pathologic</td>
</tr>
<tr>
<td>INFO/STR_STATUS</td>
<td>string list</td>
<td><a href="https://github.com/Clinical-Genomics/stranger/">Stranger</a></td>
<td>Repeat expansion status. Alternatives in [normal, pre_mutation, full_mutation]</td>
</tr>
<tr>
<td>INFO/STRAND</td>
<td>string list</td>
<td><a href="https://github.com/tjiangHIT/cuteSV">cuteSV</a></td>
<td>Strand orientation of the adjacency in BEDPE format (DEL:+-, DUP:-+, INV:++/--)</td>
</tr>
<tr>
<td>INFO/SVLEN</td>
<td>integer</td>
<td><a href="https://github.com/tjiangHIT/cuteSV">cuteSV</a>, <a href="https://github.com/fritzsedlazeck/Spectre">Spectre</a></td>
<td>Length of the SV</td>
</tr>
<tr>
<td>INFO/SVSUPPORT</td>
<td>string</td>
<td><a href="https://github.com/fritzsedlazeck/Spectre">Spectre</a></td>
<td>Indicator if a SV support was found in a provided SNFJ file</td>
</tr>
<tr>
<td>INFO/SVTYPE</td>
<td>string</td>
<td><a href="https://github.com/tjiangHIT/cuteSV">cuteSV</a>, <a href="https://github.com/fritzsedlazeck/Spectre">Spectre</a>, <a href="straglr-tsv2vcf">straglr-tsv2vcf</a></td>
<td>Type of copy number variant</td>
</tr>
<tr>
<td>INFO/SweGenMean</td>
<td>float</td>
<td><a href="https://github.com/Clinical-Genomics/stranger/">Stranger</a></td>
<td>Average number of repeat unit copies in population</td>
</tr>
<tr>
<td>INFO/SweGenStd</td>
<td>float</td>
<td><a href="https://github.com/Clinical-Genomics/stranger/">Stranger</a></td>
<td>Standard deviation of number of repeat unit copies in population</td>
</tr>
<tr>
<td>INFO/VIPC_S</td>
<td>string list</td>
<td></td>
<td>VIP decision tree classification (samples)</td>
</tr>
</tbody>
</table>
<h4 id="infocsq">INFO/CSQ<a class="headerlink" href="#infocsq" title="Permanent link">¶</a></h4>
<table>
<thead>
<tr>
<th>annotation</th>
<th>type</th>
<th>source</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td>INFO/CSQ/Allele</td>
<td>string</td>
<td>VEP</td>
<td>The variant allele used to calculate the consequence</td>
</tr>
<tr>
<td>INFO/CSQ/ALLELE_NUM</td>
<td>integer</td>
<td>VEP</td>
<td>Allele nr within the VCF file.</td>
</tr>
<tr>
<td>INFO/CSQ/ALPHSCORE</td>
<td>float</td>
<td>VEP plugin</td>
<td>AlphScore pathogenicity score for missense variants (see <a href="https://doi.org/10.1093/bioinformatics/btad280">here</a>)</td>
</tr>
<tr>
<td>INFO/CSQ/Amino_acids</td>
<td>string</td>
<td>VEP</td>
<td>Reference and variant amino acids</td>
</tr>
<tr>
<td>INFO/CSQ/apogee_Score</td>
<td>float</td>
<td>VEP plugin</td>
<td>APOGEE-2 pathogenicity score for mtDNA variants (used for protein coding genes)</td>
</tr>
<tr>
<td>INFO/CSQ/apogee_Uscore</td>
<td>float</td>
<td>VEP plugin</td>
<td>APOGEE-2 unbiased pathogenicity score mtDNA variants (used for protein coding genes)</td>
</tr>
<tr>
<td>INFO/CSQ/ASV_ACMG_class</td>
<td>string</td>
<td>VEP plugin</td>
<td>AnnotSv 'ACMG_class' output</td>
</tr>
<tr>
<td>INFO/CSQ/ASV_AnnotSV_ranking_criteria</td>
<td>string</td>
<td>VEP plugin</td>
<td>AnnotSv 'AnnotSV_ranking_criteria' output</td>
</tr>
<tr>
<td>INFO/CSQ/ASV_AnnotSV_ranking_score</td>
<td>string</td>
<td>VEP plugin</td>
<td>AnnotSv 'AnnotSV_ranking_score' output</td>
</tr>
<tr>
<td>INFO/CSQ/BIOTYPE</td>
<td>string</td>
<td>VEP</td>
<td>Biotype of transcript or regulatory feature</td>
</tr>
<tr>
<td>INFO/CSQ/CAPICE_CL</td>
<td>categorical</td>
<td>VEP plugin</td>
<td>CAPICE classification (see <a href="https://github.com/molgenis/capice">here</a>). Categories: B, LB, VUS, LP, P</td>
</tr>
<tr>
<td>INFO/CSQ/CAPICE_SC</td>
<td>float</td>
<td>VEP plugin</td>
<td>CAPICE score</td>
</tr>
<tr>
<td>INFO/CSQ/cDNA_position</td>
<td>string</td>
<td>VEP</td>
<td>Position within the cDNA</td>
</tr>
<tr>
<td>INFO/CSQ/CDS_position</td>
<td>string</td>
<td>VEP</td>
<td>Position within the coding sequence</td>
</tr>
<tr>
<td>INFO/CSQ/CHECK_REF</td>
<td>string</td>
<td>VEP</td>
<td>Reports variants where the input reference does not match the expected reference</td>
</tr>
<tr>
<td>INFO/CSQ/CLIN_SIG</td>
<td>string list</td>
<td>VEP</td>
<td>ClinVar classification(s) (do not use, see <a href="https://github.com/Ensembl/ensembl-vep/issues/1213">here</a>)</td>
</tr>
<tr>
<td>INFO/CSQ/clinVar_CLNID</td>
<td>integer list</td>
<td>VEP plugin</td>
<td>ClinVar variation identifier</td>
</tr>
<tr>
<td>INFO/CSQ/clinVar_CLNREVSTAT</td>
<td>categorical list</td>
<td>VEP plugin</td>
<td>ClinVar review status for the Variation ID. Categories: practice_guideline, reviewed_by_expert_panel, criteria_provided, _multiple_submitters, _no_conflicts, _single_submitter, _conflicting_interpretations, no_assertion_criteria_provided, no_assertion_provided</td>
</tr>
<tr>
<td>INFO/CSQ/clinVar_CLNSIG</td>
<td>string</td>
<td>VEP plugin</td>
<td>Clinical significance for this single variant; multiple values are separated by a vertical bar. Categories: Benign, Likely_benign, Uncertain_significance, Likely_pathogenic, Pathogenic, Conflicting_classifications_of_pathogenicity, Other</td>
</tr>
<tr>
<td>INFO/CSQ/clinVar_CLNSIGINCL</td>
<td>string</td>
<td>VEP plugin</td>
<td>Clinical significance for a haplotype or genotype that includes this variant. Reported as pairs of VariationID:clinical significance; multiple values are separated by a vertical bar. Categories: Benign, Likely_benign, Uncertain_significance, Likely_pathogenic, Pathogenic, Conflicting_interpretations_of_pathogenicity</td>
</tr>
<tr>
<td>INFO/CSQ/Codons</td>
<td>string</td>
<td>VEP</td>
<td>Reference and variant codon sequence</td>
</tr>
<tr>
<td>INFO/CSQ/Consequence</td>
<td>string list</td>
<td>VEP</td>
<td>Effect(s) described as Sequence Ontology term(s)</td>
</tr>
<tr>
<td>INFO/CSQ/DISTANCE</td>
<td>string</td>
<td>VEP</td>
<td>Shortest distance from variant to transcript</td>
</tr>
<tr>
<td>INFO/CSQ/existing_InFrame_oORFs</td>
<td>string</td>
<td>VEP plugin</td>
<td>The number of existing inFrame overlapping ORFs (inFrame oORF) at the 5 prime UTR</td>
</tr>
<tr>
<td>INFO/CSQ/existing_OutOfFrame_oORFs</td>
<td>string</td>
<td>VEP plugin</td>
<td>The number of existing out-of-frame overlapping ORFs (OutOfFrame oORF) at the 5 prime UTR</td>
</tr>
<tr>
<td>INFO/CSQ/existing_uORFs</td>
<td>string</td>
<td>VEP plugin</td>
<td>The number of existing uORFs with a stop codon within the 5 prime UTR</td>
</tr>
<tr>
<td>INFO/CSQ/Existing_variation</td>
<td>string list</td>
<td>VEP</td>
<td>Identifier(s) of co-located known variants</td>
</tr>
<tr>
<td>INFO/CSQ/EXON</td>
<td>string</td>
<td>VEP</td>
<td>The exon number (out of total number)</td>
</tr>
<tr>
<td>INFO/CSQ/Feature</td>
<td>string</td>
<td>VEP</td>
<td>RefSeq ID of feature</td>
</tr>
<tr>
<td>INFO/CSQ/Feature_type</td>
<td>categorical</td>
<td>VEP</td>
<td>VEP feature type. Categories: Transcript, RegulatoryFeature, MotifFeature</td>
</tr>
<tr>
<td>INFO/CSQ/FATHMM_MKL_NC</td>
<td>float</td>
<td>VEP plugin</td>
<td>The  FATHMM-MKL score for Non-Coding Single Nucleotide Variants (SNVs)</td>
</tr>
<tr>
<td>INFO/CSQ/five_prime_UTR_variant_annotation</td>
<td>string</td>
<td>VEP plugin</td>
<td>Output the annotation of a given 5 prime UTR variant</td>
</tr>
<tr>
<td>INFO/CSQ/five_prime_UTR_variant_consequence</td>
<td>string</td>
<td>VEP plugin</td>
<td>Output the variant consequences of a given 5 prime UTR variant: uAUG_gained, uAUG_lost, uSTOP_lost or uFrameshift</td>
</tr>
<tr>
<td>INFO/CSQ/FLAGS</td>
<td>string list</td>
<td>VEP</td>
<td>Transcript quality flags (cds_start_NF: CDS 5' incomplete, cds_end_NF: CDS 3' incomplete)</td>
</tr>
<tr>
<td>INFO/CSQ/GADO_PD</td>
<td>categorical</td>
<td>VEP plugin</td>
<td><a href="https://www.genenetwork.nl/gado">GADO</a> prediction for the relation between the HPO terms of the proband(s) and the gene, HC: high confidence, LC: low confidence. Categories: LC, HC</td>
</tr>
<tr>
<td>INFO/CSQ/GADO_SC</td>
<td>float</td>
<td>VEP plugin</td>
<td>The combined prioritization GADO Z-score over the HPO of the proband(s) terms for this case</td>
</tr>
<tr>
<td>INFO/CSQ/Gene</td>
<td>string</td>
<td>VEP</td>
<td>NCBI Gene ID of affected gene</td>
</tr>
<tr>
<td>INFO/CSQ/gnomAD_COV</td>
<td>float</td>
<td>VEP plugin</td>
<td>gnomAD coverage (percent of individuals in gnomAD source)</td>
</tr>
<tr>
<td>INFO/CSQ/gnomAD_AF</td>
<td>float</td>
<td>VEP plugin</td>
<td>gnomAD allele frequency</td>
</tr>
<tr>
<td>INFO/CSQ/gnomAD_FAF95</td>
<td>float</td>
<td>VEP plugin</td>
<td>gnomAD filter allele frequency (95% confidence)</td>
</tr>
<tr>
<td>INFO/CSQ/gnomAD_FAF99</td>
<td>float</td>
<td>VEP plugin</td>
<td>gnomAD filter allele frequency (99% confidence)</td>
</tr>
<tr>
<td>INFO/CSQ/gnomAD_HN</td>
<td>integer</td>
<td>VEP plugin</td>
<td>gnomAD number of homozygotes</td>
</tr>
<tr>
<td>INFO/CSQ/gnomAD_QC</td>
<td>string list</td>
<td>VEP plugin</td>
<td>gnomAD quality control filters that failed</td>
</tr>
<tr>
<td>INFO/CSQ/gnomAD_SRC</td>
<td>categorical</td>
<td>VEP plugin</td>
<td>gnomAD source (E=exomes, G=genomes, T=total)</td>
</tr>
<tr>
<td>INFO/CSQ/Grantham</td>
<td>string</td>
<td>VEP plugin</td>
<td>Grantham Matrix score - Grantham, R. Amino Acid Difference Formula to Help Explain Protein Evolution, Science 1974 Sep 6;185(4154):862-4</td>
</tr>
<tr>
<td>INFO/CSQ/HGNC_ID</td>
<td>integer</td>
<td>VEP</td>
<td>HGNC gene identifier</td>
</tr>
<tr>
<td>INFO/CSQ/HGVS_OFFSET</td>
<td>string</td>
<td>VEP</td>
<td>Indicates by how many bases the HGVS notations for this variant have been shifted</td>
</tr>
<tr>
<td>INFO/CSQ/HGVSc</td>
<td>string</td>
<td>VEP</td>
<td>HGVS nomenclature: coding DNA reference sequence</td>
</tr>
<tr>
<td>INFO/CSQ/HGVSp</td>
<td>string</td>
<td>VEP</td>
<td>HGVS nomenclature: protein reference sequence</td>
</tr>
<tr>
<td>INFO/CSQ/HIGH_INF_POS</td>
<td>string</td>
<td>VEP</td>
<td>A flag indicating if the variant falls in a high information position of a transcription factor binding profile (TFBP)</td>
</tr>
<tr>
<td>INFO/CSQ/hmtvar_DiseaseScore</td>
<td>float</td>
<td>VEP plugin</td>
<td>HmtVar disease score for mtDNA variants (used of tRNA genes)</td>
</tr>
<tr>
<td>INFO/CSQ/HPO</td>
<td>string list</td>
<td>VEP plugin</td>
<td>Human phenotype ontology term that match</td>
</tr>
<tr>
<td>INFO/CSQ/IMPACT</td>
<td>categorical</td>
<td>VEP</td>
<td>Impact as predicted by VEP. Categories: LOW, MODERATE, HIGH, MODIFIER</td>
</tr>
<tr>
<td>INFO/CSQ/IncompletePenetrance</td>
<td>string</td>
<td>VEP plugin</td>
<td>Boolean indicating if the gene is known for incomplete penetrance (1:true)</td>
</tr>
<tr>
<td>INFO/CSQ/InheritanceModesGene</td>
<td>string list</td>
<td>VEP plugin</td>
<td>List of inheritance modes for the gene</td>
</tr>
<tr>
<td>INFO/CSQ/INTRON</td>
<td>string</td>
<td>VEP</td>
<td>The intron number (out of total number)</td>
</tr>
<tr>
<td>INFO/CSQ/mitoTip_Score</td>
<td>float</td>
<td>VEP plugin</td>
<td>MitoTIP pathogenicity score for mtDNA variants (used for tRNA genes)</td>
</tr>
<tr>
<td>INFO/CSQ/mitoTip_Quartile</td>
<td>string</td>
<td>VEP plugin</td>
<td>MitoTIP pathogenicity score quartile</td>
</tr>
<tr>
<td>INFO/CSQ/MOTIF_NAME</td>
<td>string</td>
<td>VEP</td>
<td>The source and identifier of a transcription factor binding profile aligned at this position</td>
</tr>
<tr>
<td>INFO/CSQ/MOTIF_POS</td>
<td>string</td>
<td>VEP</td>
<td>The relative position of the variation in the aligned TFBP</td>
</tr>
<tr>
<td>INFO/CSQ/MOTIF_SCORE_CHANGE</td>
<td>string</td>
<td>VEP</td>
<td>The difference in motif score of the reference and variant sequences for the TFBP</td>
</tr>
<tr>
<td>INFO/CSQ/ncER</td>
<td>float</td>
<td>VEP plugin</td>
<td><a href="https://www.nature.com/articles/s41467-019-13212-3">The non-coding essential regulation (ncER)</a> score indicates if a region is likely to be essential in terms of regulation.</td>
</tr>
<tr>
<td>INFO/CSQ/PHENO</td>
<td>integer list</td>
<td>VEP</td>
<td>Indicates if existing variant is associated with a phenotype, disease or trait; multiple values correspond to multiple values in the Existing_variation field</td>
</tr>
<tr>
<td>INFO/CSQ/phyloP</td>
<td>string</td>
<td>VEP custom</td>
<td>Conservation p-values, see <a href="http://compgen.cshl.edu/phast/help-pages/phyloP.txt">here</a></td>
</tr>
<tr>
<td>INFO/CSQ/PICK</td>
<td>integer</td>
<td>VEP</td>
<td>Boolean indicating if this is the VEP picked transcript</td>
</tr>
<tr>
<td>INFO/CSQ/PolyPhen</td>
<td>float</td>
<td>VEP</td>
<td>PolyPhen score</td>
</tr>
<tr>
<td>INFO/CSQ/Protein_position</td>
<td>string</td>
<td>VEP</td>
<td>Position within the protein</td>
</tr>
<tr>
<td>INFO/CSQ/PUBMED</td>
<td>integer list</td>
<td>VEP</td>
<td>PubMed citations</td>
</tr>
<tr>
<td>INFO/CSQ/REFSEQ_MATCH</td>
<td>string</td>
<td>VEP</td>
<td>Flag indicating whether and how the RefSeq model differs from the underlying genome</td>
</tr>
<tr>
<td>INFO/CSQ/REFSEQ_OFFSET</td>
<td>string</td>
<td>VEP</td>
<td>?</td>
</tr>
<tr>
<td>INFO/CSQ/ReMM</td>
<td>float</td>
<td>VEP plugin</td>
<td>The Regulatory Mendelian Mutation (ReMM) score was created for relevance prediction of non-coding variations in the human genome in terms of Mendelian diseases.</td>
</tr>
<tr>
<td>INFO/CSQ/SIFT</td>
<td>float</td>
<td>VEP</td>
<td>SIFT score</td>
</tr>
<tr>
<td>INFO/CSQ/SOMATIC</td>
<td>integer list</td>
<td>VEP</td>
<td>Somatic status of existing variant(s); multiple values correspond to multiple values in the Existing_variation field</td>
</tr>
<tr>
<td>INFO/CSQ/SOURCE</td>
<td>string</td>
<td>VEP</td>
<td>?</td>
</tr>
<tr>
<td>INFO/CSQ/SpliceAI_pred_DP_AG</td>
<td>float</td>
<td>VEP plugin</td>
<td>SpliceAI predicted effect on splicing. Delta position for acceptor gain</td>
</tr>
<tr>
<td>INFO/CSQ/SpliceAI_pred_DP_AL</td>
<td>float</td>
<td>VEP plugin</td>
<td>SpliceAI predicted effect on splicing. Delta position for acceptor loss</td>
</tr>
<tr>
<td>INFO/CSQ/SpliceAI_pred_DP_DG</td>
<td>float</td>
<td>VEP plugin</td>
<td>SpliceAI predicted effect on splicing. Delta position for donor gain</td>
</tr>
<tr>
<td>INFO/CSQ/SpliceAI_pred_DP_DL</td>
<td>float</td>
<td>VEP plugin</td>
<td>SpliceAI predicted effect on splicing. Delta position for donor loss</td>
</tr>
<tr>
<td>INFO/CSQ/SpliceAI_pred_DS_AG</td>
<td>float</td>
<td>VEP plugin</td>
<td>SpliceAI predicted effect on splicing. Delta score for acceptor gain</td>
</tr>
<tr>
<td>INFO/CSQ/SpliceAI_pred_DS_AL</td>
<td>float</td>
<td>VEP plugin</td>
<td>SpliceAI predicted effect on splicing. Delta score for acceptor loss</td>
</tr>
<tr>
<td>INFO/CSQ/SpliceAI_pred_DS_DG</td>
<td>float</td>
<td>VEP plugin</td>
<td>SpliceAI predicted effect on splicing. Delta score for donor gain</td>
</tr>
<tr>
<td>INFO/CSQ/SpliceAI_pred_DS_DL</td>
<td>float</td>
<td>VEP plugin</td>
<td>SpliceAI predicted effect on splicing. Delta score for donor loss</td>
</tr>
<tr>
<td>INFO/CSQ/SpliceAI_pred_SYMBOL</td>
<td>string</td>
<td>VEP plugin</td>
<td>SpliceAI gene symbol</td>
</tr>
<tr>
<td>INFO/CSQ/STRAND</td>
<td>string</td>
<td>VEP</td>
<td>The DNA strand (1 or -1) on which the transcript/feature lies</td>
</tr>
<tr>
<td>INFO/CSQ/SYMBOL</td>
<td>string</td>
<td>VEP</td>
<td>Gene symbol</td>
</tr>
<tr>
<td>INFO/CSQ/SYMBOL_SOURCE</td>
<td>string</td>
<td>VEP</td>
<td>The source of the gene symbol</td>
</tr>
<tr>
<td>INFO/CSQ/TRANSCRIPTION_FACTORS</td>
<td>string</td>
<td>VEP</td>
<td>?</td>
</tr>
<tr>
<td>INFO/CSQ/VIPC</td>
<td>string</td>
<td><a href="https://github.com/molgenis/vip-decision-tree/">vip-decision-tree</a></td>
<td>VIP decision tree classification for variant effect</td>
</tr>
<tr>
<td>INFO/CSQ/VIPP</td>
<td>string list</td>
<td><a href="https://github.com/molgenis/vip-decision-tree/">vip-decision-tree</a></td>
<td>VIP decision tree path for variant effect</td>
</tr>
<tr>
<td>INFO/CSQ/VKGL</td>
<td>string</td>
<td>VEP plugin</td>
<td>?</td>
</tr>
<tr>
<td>INFO/CSQ/VKGL_CL</td>
<td>string</td>
<td>VEP plugin</td>
<td>VKGL consensus variant classification</td>
</tr>
</tbody>
</table>
<h2 id="details">Details<a class="headerlink" href="#details" title="Permanent link">¶</a></h2>
<p>VIP uses the <a href="https://github.com/Ensembl/ensembl-vep">Ensemble Effect Predictor</a> to annotate all variants with their
consequences. We use VEP with the <code>refseq</code> option for the transcripts, and with the flags for <code>sift</code> and <code>polyphen</code>
annotations enabled.</p>
<h3 id="plugins">Plugins<a class="headerlink" href="#plugins" title="Permanent link">¶</a></h3>
<p>Below we describe the other sources which we annotate using the VEP plugin framework.</p>
<h4 id="capice">CAPICE<a class="headerlink" href="#capice" title="Permanent link">¶</a></h4>
<p><a href="https://github.com/molgenis/capice">CAPICE</a> is a computational method for predicting the pathogenicity of SNVs and
InDels. It is a gradient boosting tree model trained using a variety of genomic annotations used by CADD score and
trained on the clinical significance. CAPICE performs consistently across diverse independent synthetic, and real
clinical data sets. It ourperforms the current best method in pathogenicity estimation for variants of different
molecular consequences and allele frequency.</p>
<p>We run the CAPICE application in the VIP pipeline and use a VEP plugin to annotate the VEP output with the scores from
the CAPICE output file.</p>
<h4 id="vkgl">VKGL<a class="headerlink" href="#vkgl" title="Permanent link">¶</a></h4>
<p>The datashare workgroup of VKGL has set up
a <a href="https://www.vkgl.nl/nl/diagnostiek/vkgl-datashare-database">central database</a> to enable mutual sharing of variant
classifications through a partly automatic process. An additional goal is the public sharing of these data. The
currently publicly available part of the database consists of DNA variant classifications established based on (former)
diagnostic questions.</p>
<p>We add the classifications from an export of the database and use a VEP plugin to annotate the VEP output with the
classifications from the this file.</p>
<h4 id="spliceai">SpliceAI<a class="headerlink" href="#spliceai" title="Permanent link">¶</a></h4>
<p>SpliceAI is an open-source deep learning splicing prediction algorithm that has demonstrated in the past few years its
high ability to predict splicing defects caused by DNA variations.</p>
<p>We add the scores from the available precomputed scores of SpliceAI and use a copy of the
available <a href="https://github.com/Ensembl/VEP_plugins/blob/release/109/SpliceAI.pm">VEP plugin</a> to annotate the VEP output
with the classifications from the this file.</p>
<h4 id="annotsv">AnnotSV<a class="headerlink" href="#annotsv" title="Permanent link">¶</a></h4>
<p><a href="https://lbgi.fr/AnnotSV/">AnnotSV</a> is a program for annotating and ranking structural variations from genomes of
several organisms.</p>
<p>We run the AnnotSV application in the VIP pipeline and use a VEP plugin to annotate the VEP output with the scores from
the AnnotSV output file.</p>
<h4 id="hpo">HPO<a class="headerlink" href="#hpo" title="Permanent link">¶</a></h4>
<p>A file based on the HPO <a href="http://purl.obolibrary.org/obo/hp/hpoa/phenotype_to_genes.txt">phenotype_to_genes.txt</a> is used
to annotate VEP consequences with the inheritance modes associated with the gene of this consequence.</p>
<h4 id="inheritance">Inheritance<a class="headerlink" href="#inheritance" title="Permanent link">¶</a></h4>
<p>A file based on the <a href="https://research.nhgri.nih.gov/CGD/">CGD database</a> is used to annotate VEP consequences with the
inheritance modes associated with the gene of this consequence.</p>
<h4 id="grantham">Grantham<a class="headerlink" href="#grantham" title="Permanent link">¶</a></h4>
<p>The <a href="https://www.science.org/doi/10.1126/science.185.4154.862">Grantham score</a> attempts to predict the distance between
two amino acids, in an evolutionary sense. A lower Grantham score reflects less evolutionary distance. A higher Grantham
score reflects a greater evolutionary distance.</p>
<p>We use a copy of the VEP plugin by Duarte Molha to annotate the VEP output with Grantham scores.</p>
<h4 id="gado">GADO<a class="headerlink" href="#gado" title="Permanent link">¶</a></h4>
<p><a href="https://www.nature.com/articles/s41467-019-10649-4/">GADO</a> can be used to prioritize genes based on the HPO terms of a
patient..</p>
<p>We run the GADO commandline application in the VIP pipeline and use a VEP plugin to annotate the VEP output with the
scores from the GADO output file.</p>
<h4 id="alphscore">AlphScore<a class="headerlink" href="#alphscore" title="Permanent link">¶</a></h4>
<p><a href="https://doi.org/10.5281/zenodo.6288138">AlphScore</a> is a method to predict the pathogenicity of missense variants using
features derived from AlphaFold2.</p>
<p>We add the available precomputed scores of AlphScore using a custom VEP plugin.</p>
<h4 id="ncer">ncER<a class="headerlink" href="#ncer" title="Permanent link">¶</a></h4>
<p><a href="https://www.nature.com/articles/s41467-019-13212-3">The non-coding essential regulation (ncER)</a> score indicates if a
region is likely to be essential in terms of regulation.
The ncER file VIP uses is the version provided by GREEN-VARAN (https://github.com/edg1983/GREEN-VARAN) on
Zenodo: https://zenodo.org/records/5636163
If overlapping regions are encountered (which can occur in liftovered resources) the highest score is annotated.</p>
<h4 id="remm">ReMM<a class="headerlink" href="#remm" title="Permanent link">¶</a></h4>
<p><a href="https://charite.github.io/software-remm-score.html">The Regulatory Mendelian Mutation (ReMM)</a> score was created for
relevance prediction of non-coding variations (SNVs and small InDels) in the human genome (hg19) in terms of Mendelian
diseases. The VEP plugin is build on top of the GREEN-DB dataset (GRCh38) for ReMM
scores: https://zenodo.org/records/3955933
If overlapping regions are encountered (which can occur in liftovered resources) the highest score is annotated.</p>
<h4 id="fathmm-mkl">FATHMM-MKL<a class="headerlink" href="#fathmm-mkl" title="Permanent link">¶</a></h4>
<p><a href="https://fathmm.biocompute.org.uk/fathmmMKL.htm">FATHMM-MKL</a> predicts the Functional Consequences of Coding and
Non-Coding Single Nucleotide Variants (SNVs)
This plugin annotates non-coding scores only, and is build on top of the GREEN-DB dataset (GRCh38) for FATHMM-MKL non
coding scores: https://zenodo.org/records/3981121</p>
<h4 id="green-db-constraint-scores">GREEN-DB constraint scores<a class="headerlink" href="#green-db-constraint-scores" title="Permanent link">¶</a></h4>
<p><a href="https://doi.org/10.1101/2020.09.17.301960">GREEN-DB</a> is a comprehensive collection of 2.4 million regulatory
elements in the human genome collected from previously published databases, high-throughput screenings and functional
studies.
This plugin annotates the constrain scores only, and is build on top of the GREEN-DB bed files (
GRCh38): https://zenodo.org/records/5636209
GREEN-DB constrains scores are annotated per region type: enhancers, promotors, bivalent, insulators, silencers. If
multiple regions of the same type overlap, VIP annotates the highest constraint score.</p>
<h4 id="apogee-2">APOGEE 2<a class="headerlink" href="#apogee-2" title="Permanent link">¶</a></h4>
<p><a href="https://doi.org/10.1038/s41467-023-40797-7">APOGEE-2</a> predicts pathogenicity scores for missense mitochondrial variants.
A custom APOGEE VEP plugin annotates mitochondrial variants with computed <a href="https://mitimpact.mcb2lab.org">APOGEE scores</a>.</p>
<h4 id="mitotip">MitoTIP<a class="headerlink" href="#mitotip" title="Permanent link">¶</a></h4>
<p><a href="https://doi.org/10.1371/journal.pcbi.1005867">MitoTIP</a> predicts pathogenicity scores for mtDNA tRNA variants.
A custom MitoTIP VEP plugin annotates mitochondrial variants with computed <a href="https://www.mitomap.org/MITOMAP/MitoTipScores">mitoTip scores</a> and their corresponding quartiles (Q1-Q4).</p>
<h4 id="hmtvar">HmtVar<a class="headerlink" href="#hmtvar" title="Permanent link">¶</a></h4>
<p><a href="https://doi.org/10.1093/nar/gky1024">HmtVar</a> contains disease scores for mtDNA SNVs.
A custom HmtVar VEP plugin annotates mitochondrial tRNA variants with computed disease scores.</p>
</div>
</div><footer>
<div aria-label="Footer Navigation" class="rst-footer-buttons" role="navigation">
<a class="btn btn-neutral float-left" href="../../examples/reanalysis/" title="Reanalysis"><span class="icon icon-circle-arrow-left"></span> Previous</a>
<a class="btn btn-neutral float-right" href="../classification_trees/" title="Classification trees">Next <span class="icon icon-circle-arrow-right"></span></a>
</div>
<hr/>
<div role="contentinfo">
<!-- Copyright etc -->
<p>Copyright © 2026 University Medical Center Groningen All rights reserved</p>
</div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
</div>
</div>
</section>
</div>
<div aria-label="Versions" class="rst-versions" role="note">
<span class="rst-current-version" data-toggle="rst-current-version">
<span>
<a class="fa fa-github" href="https://github.com/molgenis/vip/" style="color: #fcfcfc"> GitHub</a>
</span>
<span><a href="../../examples/reanalysis/" style="color: #fcfcfc">« Previous</a></span>
<span><a href="../classification_trees/" style="color: #fcfcfc">Next »</a></span>
</span>
</div>
<script src="../../js/jquery-3.6.0.min.js"></script>
<script>var base_url = "../..";</script>
<script src="../../js/theme_extra.js"></script>
<script src="../../js/theme.js"></script>
<script src="../../js/mermaid.min.js"></script>
<script src="../../search/main.js"></script>
<script src="../../assets/javascripts/panzoom.min.js"></script>
<script src="../../assets/javascripts/zoompan.js"></script>
<script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>
</body>
</html>
